README
NAME
    Map::Tube::Samara - Interface to the Samara Metro Map.

SYNOPSIS
     use Map::Tube::Samara;
     my $obj = Map::Tube::Samara->new;
     my $routes_ar = $obj->get_all_routes($from, $to);
     my $line = $obj->get_line_by_id($line_id);
     my $line = $obj->get_line_by_name($line_name);
     my $lines_ar = $obj->get_lines;
     my $station = $obj->get_node_by_id($station_id);
     my $station = $obj->get_node_by_name($station_name);
     my $route = $obj->get_shortest_route($from, $to);
     my $stations_ar = $obj->get_stations($line);
     my $metro_name = $obj->name;
     my $xml_file = $obj->xml;

DESCRIPTION
    It currently provides functionality to find the shortest route between
    the two given nodes.

    For more information about Samara Map, click here
    <https://en.wikipedia.org/wiki/Samara_Metro>.

METHODS
    "new()"
             Constructor.

    "get_all_routes($from, $to)" [EXPERIMENTAL]
             Get all routes from station to station.
             Returns reference to array with Map::Tube::Route objects.

    "get_line_by_id($line_id)"
             Get line object defined by id.
             Returns Map::Tube::Line object.

    "get_line_by_name($line_name)"
             Get line object defined by name.
             Returns Map::Tube::Line object.

    "get_lines()"
             Get lines in metro map.
             Returns reference to unsorted array with Map::Tube::Line objects.

    "get_node_by_id($station_id)"
             Get station node by id.
             Returns Map::Tube::Node object.

    "get_node_by_name($station_name)"
             Get station node by name.
             Returns Map::Tube::Node object.

    "get_shortest_route($from, $to)"
             Get shortest route between $from and $to node names. Node names in $from and $to are case insensitive.
             Returns Map::Tube::Route object.

    "get_stations($line)"
             Get list of stations for concrete metro line.
             Returns reference to array with Map::Tube::Node objects.

    "name()"
             Get metro name.
             Returns string with metro name.

    "xml()"
             Get XML specification of Samara metro.
             Returns string with XML.

EXAMPLE1
     # Pragmas.
     use strict;
     use warnings;

     # Modules.
     use Encode qw(decode_utf8 encode_utf8);
     use Map::Tube::Samara;

     # Object.
     my $obj = Map::Tube::Samara->new;

     # Get route.
     my $route = $obj->get_shortest_route(decode_utf8('Гагаринская'), decode_utf8('Безымянка'));

     # Print out type.
     print "Route: ".encode_utf8($route)."\n";

     # Output:
     # Route: Гагаринская (Первая линия), Спортивная (Первая линия), Советская (Первая линия), Победа (Первая линия), Безымянка (Первая линия)

EXAMPLE2
     # Pragmas.
     use strict;
     use warnings;

     # Modules.
     use Map::Tube::Samara;

     # Object.
     my $obj = Map::Tube::Samara->new;

     # Get XML file.
     my $xml_file = $obj->xml;

     # Print out XML file.
     print "XML file: $xml_file\n";

     # Output like:
     # XML file: .*/samara-map.xml

EXAMPLE3
     # Pragmas.
     use strict;
     use warnings;

     # Modules.
     use Map::Tube::GraphViz;
     use Map::Tube::GraphViz::Utils qw(node_color_without_label);
     use Map::Tube::Samara;

     # Object.
     my $obj = Map::Tube::Samara->new;

     # GraphViz object.
     my $g = Map::Tube::GraphViz->new(
             'callback_node' => \&node_color_without_label,
             'driver' => 'neato',
             'tube' => $obj,
     );

     # Get graph to file.
     $g->graph('Samara.png');

     # Print file.
     system "ls -l Samara.png";

     # Output like:
     # -rw-r--r-- 1 skim skim 28389 Jan  8 21:11 Samara.png

EXAMPLE4
     # Pragmas.
     use strict;
     use warnings;

     # Modules.
     use Encode qw(encode_utf8);
     use Map::Tube::Samara;

     # Object.
     my $obj = Map::Tube::Samara->new;

     # Get lines.
     my $lines_ar = $obj->get_lines;

     # Print out.
     map { print encode_utf8($_->name)."\n"; } sort @{$lines_ar};

     # Output:
     # Первая линия

EXAMPLE5
     # Pragmas.
     use strict;
     use warnings;

     # Modules.
     use Encode qw(decode_utf8 encode_utf8);
     use Map::Tube::Samara;

     # Arguments.
     if (@ARGV < 1) {
             print STDERR "Usage: $0 line\n";
             exit 1;
     }
     my $line = decode_utf8($ARGV[0]);

     # Object.
     my $obj = Map::Tube::Samara->new;

     # Get stations for line.
     my $stations_ar = $obj->get_stations($line);

     # Print out.
     map { print encode_utf8($_->name)."\n"; } @{$stations_ar};

     # Output:
     # Usage: __PROG__ line

     # Output with 'foo' argument.
     # Map::Tube::get_stations(): ERROR: Invalid Line Name [foo]. (status: 105) file __PROG__ on line __LINE__

     # Output with 'Первая линия' argument.
     # Российская
     # Московская
     # Гагаринская
     # Спортивная
     # Советская
     # Победа
     # Безымянка
     # Кировская
     # Юнгородок

DEPENDENCIES
    File::Share, Map::Tube, Moo, namespace::clean.

SEE ALSO
    Map::Tube
        Core library as Role (Moo) to process map data.

    Task::Map::Tube
        Install the Map::Tube modules.

    Task::Map::Tube::Metro
        Install the Map::Tube concrete metro modules.

REPOSITORY
    <https://github.com/tupinek/Map-Tube-Samara>

AUTHOR
    Michal Josef Špaček <mailto:skim@cpan.org>

    <http://skim.cz>

LICENSE AND COPYRIGHT
     © 2014-2017 Michal Josef Špaček
     Artistic License
     BSD 2-Clause License

VERSION
    0.08



Changes
0.08
 - Fix Changelog file.

0.07 2017-11-22T08:40:52+01:00
 - Add test for ok_map_functions().
 - Fix Makefile.PL for new version of Perl and no default »use lib '.'«.
 - Remove bad test.
 - Update Module::Install to 1.18 version.
 - Update author name.
 - Update copyright years to 2017.
 - Update to Map::Tube 3.41.

0.06 2015-11-05T10:23:20+01:00
 - Add EXAMPLE5 section to doc.
 - Add commented line 3 to XML map.
 - Fix commented line 2 in XML map.
 - Implement station order.
 - Rewrite XML map to new version with line IDs.

0.05 2015-10-16T12:43:32+02:00
 - Add tests for get_line_by_id(), get_line_by_name(), get_lines(), and
   get_stations() methods.
 - Fix test dependencies.

0.04 2015-10-16T11:49:15+02:00
 - Add lines to XML document.
 - Update to Map::Tube 3.06.

0.03 2015-10-13T10:31:14+02:00
 - Add <?xml> header to XML document.
 - Add 04-map.t test.
 - Add EXAMPLE4 to doc.
 - Add get_line_by_id() and get_line_by_name() methods to doc.
 - Add tests for get_node_by_id() and get_node_by_name() methods.
 - Improve description for get_lines() in doc.
 - Move author tests to xt/ directory.
 - Rewrite SEE ALSO section to new style.
 - Update Module::Install to 1.16 version.

0.02 2015-01-09T07:35:58+01:00
 - Add EXAMPLE3 section to doc.
 - Add Map::Tube::GraphViz, Map::Tube::Text::Table, Task::Map::Tube,
   Map::Tube::Bucharest, Map::Tube::Budapest, Map::Tube::Dnipropetrovsk,
   Map::Tube::Kazan, Map::Tube::Kharkiv, Map::Tube::Nanjing,
   Map::Tube::Novosibirsk, Map::Tube::SaintPetersburg, Map::Tube::Tbilisi,
   Map::Tube::Vienna and Map::Tube::Yekaterinburg to SEE ALSO section in
   doc.
 - Add get_lines() method to doc.
 - Add get_stations() method to doc.
 - Remove executable bit from samara-map.xml file.
 - Remove obsolete utf8 pragma from EXAMPLE2 section in doc.
 - Update copyright years to 2015.

0.01 2014-12-09T10:30:44+01:00
 - First version


Makefile.PL
# Pragmas.
use lib 'inc';
use strict;
use warnings;

# Modules.
use inc::Module::Install;

# Definition.
abstract 'Interface to the Samara Metro Map.';
author 'Michal Josef Spacek <skim@cpan.org>';
author_requires 'English' => 0;
author_requires 'File::Object' => 0.08;
author_requires 'Test::More' => 0;
author_requires 'Test::NoWarnings' => 0;
author_requires 'Test::Pod' => 0;
author_requires 'Test::Pod::Coverage' => 0;
install_share 'share';
license 'bsd';
license 'artistic_2';
name 'Map-Tube-Samara';
readme_from 'Samara.pm';
recursive_author_tests('xt');
requires 'File::Share' => 0;
requires 'Map::Tube' => '3.41';
requires 'Moo' => 0;
requires 'namespace::clean' => 0;
requires 'perl' => '5.8.0';
resources 'bugtracker' => 'https://rt.cpan.org/Public/Dist/Display.html'.
	'?Name=Map-Tube-Samara';
resources 'homepage' => 'https://github.com/tupinek/Map-Tube-Samara';
resources 'repository' => 'git://github.com/tupinek/Map-Tube-Samara';
test_requires 'Encode' => 0;
test_requires 'English' => 0;
test_requires 'Test::Map::Tube' => 0.35;
test_requires 'Test::More' => 0;
test_requires 'Test::NoWarnings' => 0;
tests_recursive;
version '0.08';

# Run.
WriteAll();


Samara.pm
package Map::Tube::Samara;

# Pragmas.
use strict;
use warnings;
use 5.006;

# Modules.
use File::Share ':all';
use Moo;
use namespace::clean;

# Version.
our $VERSION = 0.08;

# Get XML.
has xml => (
	'is' => 'ro',
	'default' => sub {
		return dist_file('Map-Tube-Samara', 'samara-map.xml');
	},
);

with 'Map::Tube';

1;

__END__

=encoding utf8

=head1 NAME

Map::Tube::Samara - Interface to the Samara Metro Map.

=head1 SYNOPSIS

 use Map::Tube::Samara;
 my $obj = Map::Tube::Samara->new;
 my $routes_ar = $obj->get_all_routes($from, $to);
 my $line = $obj->get_line_by_id($line_id);
 my $line = $obj->get_line_by_name($line_name);
 my $lines_ar = $obj->get_lines;
 my $station = $obj->get_node_by_id($station_id);
 my $station = $obj->get_node_by_name($station_name);
 my $route = $obj->get_shortest_route($from, $to);
 my $stations_ar = $obj->get_stations($line);
 my $metro_name = $obj->name;
 my $xml_file = $obj->xml;

=head1 DESCRIPTION

It currently provides functionality to find the shortest route between the two
given nodes.

For more information about Samara Map, click L<here|https://en.wikipedia.org/wiki/Samara_Metro>.

=head1 METHODS

=over 8

=item C<new()>

 Constructor.

=item C<get_all_routes($from, $to)> [EXPERIMENTAL]

 Get all routes from station to station.
 Returns reference to array with Map::Tube::Route objects.

=item C<get_line_by_id($line_id)>

 Get line object defined by id.
 Returns Map::Tube::Line object.

=item C<get_line_by_name($line_name)>

 Get line object defined by name.
 Returns Map::Tube::Line object.

=item C<get_lines()>

 Get lines in metro map.
 Returns reference to unsorted array with Map::Tube::Line objects.

=item C<get_node_by_id($station_id)>

 Get station node by id.
 Returns Map::Tube::Node object.

=item C<get_node_by_name($station_name)>

 Get station node by name.
 Returns Map::Tube::Node object.

=item C<get_shortest_route($from, $to)>

 Get shortest route between $from and $to node names. Node names in $from and $to are case insensitive.
 Returns Map::Tube::Route object.

=item C<get_stations($line)>

 Get list of stations for concrete metro line.
 Returns reference to array with Map::Tube::Node objects.

=item C<name()>

 Get metro name.
 Returns string with metro name.

=item C<xml()>

 Get XML specification of Samara metro.
 Returns string with XML.

=back

=head1 EXAMPLE1

 # Pragmas.
 use strict;
 use warnings;

 # Modules.
 use Encode qw(decode_utf8 encode_utf8);
 use Map::Tube::Samara;

 # Object.
 my $obj = Map::Tube::Samara->new;

 # Get route.
 my $route = $obj->get_shortest_route(decode_utf8('Гагаринская'), decode_utf8('Безымянка'));

 # Print out type.
 print "Route: ".encode_utf8($route)."\n";

 # Output:
 # Route: Гагаринская (Первая линия), Спортивная (Первая линия), Советская (Первая линия), Победа (Первая линия), Безымянка (Первая линия)

=head1 EXAMPLE2

 # Pragmas.
 use strict;
 use warnings;

 # Modules.
 use Map::Tube::Samara;

 # Object.
 my $obj = Map::Tube::Samara->new;

 # Get XML file.
 my $xml_file = $obj->xml;

 # Print out XML file.
 print "XML file: $xml_file\n";

 # Output like:
 # XML file: .*/samara-map.xml

=head1 EXAMPLE3

 # Pragmas.
 use strict;
 use warnings;

 # Modules.
 use Map::Tube::GraphViz;
 use Map::Tube::GraphViz::Utils qw(node_color_without_label);
 use Map::Tube::Samara;

 # Object.
 my $obj = Map::Tube::Samara->new;

 # GraphViz object.
 my $g = Map::Tube::GraphViz->new(
         'callback_node' => \&node_color_without_label,
         'driver' => 'neato',
         'tube' => $obj,
 );

 # Get graph to file.
 $g->graph('Samara.png');

 # Print file.
 system "ls -l Samara.png";

 # Output like:
 # -rw-r--r-- 1 skim skim 28389 Jan  8 21:11 Samara.png

=begin html

<a href="https://raw.githubusercontent.com/tupinek/Map-Tube-Samara/master/images/ex3.png">
  <img src="https://raw.githubusercontent.com/tupinek/Map-Tube-Samara/master/images/ex3.png" alt="Самарский метрополитен" width="300px" height="300px" />
</a>

=end html

=head1 EXAMPLE4

 # Pragmas.
 use strict;
 use warnings;

 # Modules.
 use Encode qw(encode_utf8);
 use Map::Tube::Samara;

 # Object.
 my $obj = Map::Tube::Samara->new;

 # Get lines.
 my $lines_ar = $obj->get_lines;

 # Print out.
 map { print encode_utf8($_->name)."\n"; } sort @{$lines_ar};

 # Output:
 # Первая линия

=head1 EXAMPLE5

 # Pragmas.
 use strict;
 use warnings;

 # Modules.
 use Encode qw(decode_utf8 encode_utf8);
 use Map::Tube::Samara;

 # Arguments.
 if (@ARGV < 1) {
         print STDERR "Usage: $0 line\n";
         exit 1;
 }
 my $line = decode_utf8($ARGV[0]);

 # Object.
 my $obj = Map::Tube::Samara->new;

 # Get stations for line.
 my $stations_ar = $obj->get_stations($line);

 # Print out.
 map { print encode_utf8($_->name)."\n"; } @{$stations_ar};

 # Output:
 # Usage: __PROG__ line

 # Output with 'foo' argument.
 # Map::Tube::get_stations(): ERROR: Invalid Line Name [foo]. (status: 105) file __PROG__ on line __LINE__

 # Output with 'Первая линия' argument.
 # Российская
 # Московская
 # Гагаринская
 # Спортивная
 # Советская
 # Победа
 # Безымянка
 # Кировская
 # Юнгородок

=head1 DEPENDENCIES

L<File::Share>,
L<Map::Tube>,
L<Moo>,
L<namespace::clean>.

=head1 SEE ALSO

=over

=item L<Map::Tube>

Core library as Role (Moo) to process map data.

=item L<Task::Map::Tube>

Install the Map::Tube modules.

=item L<Task::Map::Tube::Metro>

Install the Map::Tube concrete metro modules.

=back

=head1 REPOSITORY

L<https://github.com/tupinek/Map-Tube-Samara>

=head1 AUTHOR

Michal Josef Špaček L<mailto:skim@cpan.org>

L<http://skim.cz>

=head1 LICENSE AND COPYRIGHT

 © 2014-2017 Michal Josef Špaček
 Artistic License
 BSD 2-Clause License

=head1 VERSION

0.08

=cut
